### PR TITLE
Preserve property order in environment editor

### LINKED ISSUE #1046: [Feature Request] Sort environment variables
- Insomnia Version: 5.16.6
- Operating System:Ubuntu 18.04
## Details

The variables in "Manage Environments" are automatically ordered alphabetically. While this is good some of the time, sometimes I'd like to logically group certain variables together and in a certain order.

It would be nice to have this optional - perhaps the alphabetical sorting could be a (default on) option?

### PR DESCRIPTION
**Pull request by develohpanda**
Closes #1046.

image

Added an option to keep property order by consuming [json-order](https://github.com/develohpanda/json-order). This is opt-in, thus **disabled** by default. This behavior is available to both the workspace environment editor, and the request group override environment editor.

Field names cannot begin by `$` or contain a `.` (as part of the `nedb` [documentation](https://github.com/louischatriot/nedb/blob/aa3302e5dcefa471431901a8f3fe63e532a5a7b2/README.md#inserting-documents)). 

As a result, the map generation uses `&` as the prefix and `~|` as the separator. These two options are parameters in the `parse` and `stringify` methods. An example map is shown below.

```js
const map = {
  "&": ["a"],
  "&~|a": ["z", "y"],
  "&~|a~|z": ["b"],
  "&~|a~|y~|2": ["d", "c"]
}
```

### CODE REVIEW

File: packages/insomnia-app/app/ui/components/modals/workspace-environments-edit-modal.js [hunk 1]
```diff
@@ -441,39 +451,39 @@ class WorkspaceEnvironmentsEditModal extends React.PureComponent<Props, State> {
...
                     value={activeEnvironment}
-                    onClick={this._handleUnsetColor}
-                    disabled={!activeEnvironment.color}>
-                    <i className="fa fa-minus-circle" />
-                    Unset Color
-                  </DropdownItem>
-                </Dropdown>
-              ) : null}
-
-              {activeEnvironment && rootEnvironment !== activeEnvironment ? (
```
Comment (develohpanda) [inline packages/insomnia-app/app/ui/components/modals/workspace-environments-edit-modal.js]: This condition is identical to the one above, so I pull both components into a parent `React.Fragment`. Easier to see with a "no whitespace" diff.

**Comment by develohpanda** [comment]
I'm not sure I added the dependency correctly though - should `json-order` be under `packedDependencies` or `dependencies`?

**Comment by gschier** [comment]
> I'm not sure I added the dependency correctly - should `json-order` be under `packedDependencies` or `dependencies`?

It should be in both. `packedDependencies` is a special insomnia-specific property that the build script uses to figure out whether the dependency should be bundled in the Webpack output or left as an external node module. In general, most should be bundled, but some modules don't work like that so need to be left alone.

**Comment by develohpanda** [comment]
Hmmm - I'm not sure why the CI builds are failing. I added to the packed dep but it still failed. Surely I've missed a step somewhere while adding the package 🤔 

image

**Comment by gschier** [comment]
Ah, @develohpanda it's because you need to add a Flow definition to the `flow-typed` folder. I wish there were a way to have a better error message when that happens.

Here's a good example of one to take a look at: https://github.com/getinsomnia/insomnia/blob/639d0cc33b0d455f43f449da1cbbe42dffa443cc/packages/insomnia-app/flow-typed/mkdirp.js

File: packages/insomnia-app/app/models/environment.js [hunk 2]
```diff
@@ -26,6 +27,7 @@ export function init() {
   return {
     name: 'New Environment',
     data: {},
+    propertyMap: null,
```
Comment (gschier) [inline packages/insomnia-app/app/models/environment.js]: Can we rename this to `dataPropertyOrder` in order to better signify what it's for?

File: packages/insomnia-app/app/models/request-group.js [hunk 3]
```diff
@@ -22,6 +23,7 @@ export function init() {
     name: 'New Folder',
     description: '',
     environment: {},
+    propertyMap: null,
```
Comment (gschier) [inline packages/insomnia-app/app/models/request-group.js]: Same with this one. `environmentPropertyOrder` seems more descriptive. I'm open to other suggestions though.

File: packages/insomnia-app/app/ui/components/editors/environment-editor.js [hunk 4]
```diff
@@ -107,16 +134,32 @@ class EnvironmentEditor extends React.PureComponent<Props, State> {
...
           render={render}
           getRenderContext={getRenderContext}
           mode="application/json"
           {...props}
         />
-        {error && <p className="notice error margin">{error}</p>}
-        {!error && warning && <p className="notice warning margin">{warning}</p>}
+        {error && <p className="notice error margin-x margin-y-sm">{error}</p>}
+        {!error && warning && <p className="notice warning margin-x margin-y-sm">{warning}</p>}
+        <div className="form-control margin-x margin-y-sm">
```
Comment (gschier) [inline packages/insomnia-app/app/ui/components/editors/environment-editor.js]: My feeling is that we probably don't need to make this an option. It seems like maintaining order is expected no matter what. Thoughts?

File: packages/insomnia-app/app/ui/components/editors/environment-editor.js [hunk 5]
```diff
@@ -69,11 +78,29 @@ class EnvironmentEditor extends React.PureComponent<Props, State> {
     this._editor = n;
   }
 
-  getValue() {
+  _updateMaintainOrderBoolean(checked: boolean) {
+    this.setState({ maintainOrder: !this.state.maintainOrder }, () => {
+      this.props.didChange();
```
Comment (gschier) [inline packages/insomnia-app/app/ui/components/editors/environment-editor.js]: Nice! I always forget setState has a callback param 👍

File: packages/insomnia-app/app/ui/components/modals/environment-edit-modal.js [hunk 6]
```diff
@@ -30,10 +30,20 @@ class EnvironmentEditModal extends PureComponent {
       return;
     }
 
-    const environment = this._envEditor.getValue();
+    let patch;
```
Comment (gschier) [inline packages/insomnia-app/app/ui/components/modals/environment-edit-modal.js]: Curious why you included the try/catch when there wasn't one before. Did you find an error case that wasn't handled?

File: packages/insomnia-app/app/ui/components/modals/workspace-environments-edit-modal.js [hunk 7]
```diff
@@ -342,9 +343,13 @@ class WorkspaceEnvironmentsEditModal extends React.PureComponent<Props, State> {
       return;
     }
 
-    let data;
+    let patch;
     try {
```
Comment (gschier) [inline packages/insomnia-app/app/ui/components/modals/workspace-environments-edit-modal.js]: Oh I see. Same thing here. I just checked the `getValue` function and didn't realize it was doing JSON parsing. Makes sense now 👍

File: packages/insomnia-app/app/ui/components/modals/workspace-environments-edit-modal.js (hunk 1 above)
Comment (gschier) [reply packages/insomnia-app/app/ui/components/modals/workspace-environments-edit-modal.js]: Nice! I bet that old code was written before `React.Fragment` existed 👀

File: packages/insomnia-app/app/ui/css/layout/base.less [hunk 8]
```diff
@@ -554,6 +554,16 @@ i.fa.fa--fixed-width {
   margin: @padding-md;
 }
 
+.margin-x {
```
Comment (gschier) [inline packages/insomnia-app/app/ui/css/layout/base.less]: Nice 👍 Exactly the names I would have chosen too

File: packages/insomnia-app/app/ui/components/editors/environment-editor.js (hunk 4 above)
Comment (develohpanda) [reply packages/insomnia-app/app/ui/components/editors/environment-editor.js]: Yeah fair - initial feature request mentioned for it to be optional but I don't see a particularly compelling use-case for not keeping order.

This checkbox is powered by the fact that a propertyOrder object exists or not - so its not a lot of extra work to remove. :)

**Comment by develohpanda** [comment]
@gschier ready for review again :)

**Pull request merged**